linode/firewall/framework_models.go

```diff
@@ -0,0 +1,168 @@
+	Status         types.String `tfsdk:"status"`
+}
+
+func (data *FirewallModel) parseComputedAttributes(
+	ctx context.Context,
+	firewall *linodego.Firewall,
+	rules *linodego.FirewallRuleSet,
+	devices []linodego.FirewallDevice,
+) diag.Diagnostics {
+	data.ID = types.Int64Value(int64(firewall.ID))
+	data.Disabled = types.BoolValue(firewall.Status == linodego.FirewallDisabled)
+	data.Status = types.StringValue(string(firewall.Status))
+
+	linodes, diag := types.SetValueFrom(ctx, types.Int64Type, parseFirewallLinodes(devices))
+	if diag.HasError() {
+		return diag
+	}
+	data.Linodes = linodes
+
+	firewallDevices, diag := parseFirewallDevices(devices)
+	if diag.HasError() {
+		return diag
+	}
+	data.Devices = *firewallDevices
+
+	return nil
+}
+
+func (data *FirewallModel) parseNonComputedAttributes(
+	ctx context.Context,
+	firewall *linodego.Firewall,
+	rules *linodego.FirewallRuleSet,
+	devices []linodego.FirewallDevice,
+) diag.Diagnostics {
+	tags, diag := types.SetValueFrom(ctx, types.StringType, firewall.Tags)
+	if diag.HasError() {
+		return diag
+	}
+	data.Tags = tags
+	data.InboundPolicy = types.StringValue(rules.InboundPolicy)
+	data.OutboundPolicy = types.StringValue(rules.OutboundPolicy)
+	data.Label = types.StringValue(firewall.Label)
+
+	if rules.Inbound != nil {
+		inBound, diag := parseFirewallRules(ctx, rules.Inbound)
+		if diag.HasError() {
+			return diag
+		}
+		data.Inbound = *inBound
+	}
+
+	if rules.Outbound != nil {
+		outBound, diag := parseFirewallRules(ctx, rules.Outbound)
+		if diag.HasError() {
+			return diag
+		}
+		data.Outbound = *outBound
+	}
+
+	return nil
+}
+
+func parseFirewallRules(
+	ctx context.Context,
+	rules []linodego.FirewallRule,
+) (*basetypes.ListValue, diag.Diagnostics) {
+	specs := make([]attr.Value, len(rules))
+
+	for i, rule := range rules {
+		spec := make(map[string]attr.Value)
+		spec["action"] = types.StringValue(rule.Action)
+		spec["label"] = types.StringValue(rule.Label)
+		spec["protocol"] = types.StringValue(string(rule.Protocol))
+		spec["ports"] = types.StringValue(rule.Ports)
+		ipv4, diag := types.ListValueFrom(ctx, types.StringType, rule.Addresses.IPv4)
+		if diag.HasError() {
+			return nil, diag
+		}
+		spec["ipv4"] = ipv4
+
+		ipv6, diag := types.ListValueFrom(ctx, types.StringType, rule.Addresses.IPv6)
+		if diag.HasError() {
+			return nil, diag
+		}
+		spec["ipv6"] = ipv6
+
+		obj, diag := types.ObjectValue(datasourceRuleObjectType.AttrTypes, spec)
+		if diag.HasError() {
+			return nil, diag
+		}
+
+		specs[i] = obj
+	}
+
+	result, diag := basetypes.NewListValue(
+		datasourceRuleObjectType,
+		specs,
+	)
+	if diag.HasError() {
+		return nil, diag
+	}
+
+	return &result, nil
+}
+
+func parseFirewallLinodes(devices []linodego.FirewallDevice) []int {
+	linodes := make([]int, 0, len(devices))
+	for _, device := range devices {
+		if device.Entity.Type == linodego.FirewallDeviceLinode {
+			linodes = append(linodes, device.Entity.ID)
+		}
+	}
+	return linodes
+}
```

Comment:
Since this function wasn't changed, can we still use `flattenFirewallLinodes` instead of defining it again?

Reply:
I'm thinking we may want to remove the `helpers.go` entirely after migrating both, so I copied it here. Leave the `flattenFirewallLinodes` function in the helpers for now to make sure resource still can work.

Reply:
Since firewall module contains both resource and data source, I think we may keep the helpers that could be used to share some common functions?

Reply:
My understanding is the functions in the helpers.go currently can all be modified and moved to `framework_models` later when doing resource migration.

Reply:
Ah, yes. You are right.